### ISSUE TITLE
[Bug]: Conky not visible when XWayland is started after/by it

### ISSUE THREAD

**Issue by rochfort1** [brndnmtthws/conky#1087]
**Issue**

With GNOME 40.0.0, if you are using the default logon option, GNOME, conky does not show on the desktop after it has been started. (However, conky does show if you log on to GNOME with the GNOME on Xorg option). After starting GNOME, if you inspect **System Monitor**, a conky process is listed but conky does not show on the desktop.
It seems that Conky uses Xwayland's on-demand feature but does not wait for Xwayland to start.

**Information**
The following two tests reveal the bug. The first test shows that conky does not show on the desktop if it is called before Xwayland has started. The second test shows that conky shows on the desktop if it is called when Xwayland is already running.

_First test_
1. Make sure that no software requiring Xwayland is set to automatically start when you log on to GNOME.
2. Log off GNOME and then log on again.
3. Open **System Monitor** and verify that Xwayland is not listed.
4. Open **Terminal** and start conky.
5. In **System Monitor**, verify that both conky and Xwayland are listed.
6. Verify that conky does not show on the desktop.

_Second test_
1. Log off GNOME and then log on again.
2. Open **System Monitor** and verify that Xwayland is not listed.
3. Open **Terminal** and start xeyes.
4. In **System Monitor**, verify that both xeyes and Xwayland are listed.
5. Verify that xeyes shows on the desktop.
6. Close xeyes and, in **System Monitor**, verify that xeyes is not listed and that Xwaylend is listed.
7. In **Terminal**, start conky.
8. In **System Monitor**, verify that conky is listed and, on the desktop, verify that conky shows.

Notes
(a) With the second test, you must log off and then log on again. If, at the end of the first test, you simply used **System Monitor** to end the process for both conky and Xwayland and then started conky, conky will show on the desktop. It's as though Xwayland hangs about after **End Process** has been initiated.

(b) On GNOME 3.38.4, conky always started and showed on the desktop. However I noticed that on GNOME 3.38.4, Xwayland is always running.

(c) The first and second tests performed the same on Arch Linux and Fedora 34. Both used GNOME 40.0.0.

**Comment by barmadrid** [comment]
Same issue on Arch Linux with latest conky-git (commit d966e8522a327842ff2297f7521bfab78e17941d).

**Comment by chromer030** [comment]
Isn't time for conky to run on native wayland ?

**Comment by jason-mehmel** [comment]
I've been running conky fine for the last few weeks with the below environment.

I just did an upgrade and restart in Ubuntu 20.04 with Wayland yesterday, and when I came back, Conky wasn't working!

Running GNOME 3.36.8, 
Windowing system Wayland.

Did I just catch up to a new problem with conky and Wayland?

**Comment by Caellian** [comment]
I'm reopening this issue but marking it as low priority because conky should indeed run natively on wayland.

It _is_ however a bug because X11 code should handle these conditions properly. What I'm assuming is happening is that both XWayland and Conky are initialized concurrently and some necessary thing isn't made available to conky yet. This can occur even if XWayland isn't used - when conky is started from the same script as X11 WM/DE. So it's a thing that should be handled properly.

Among bugs, this is very niche, but during triage I've personally instructed several people how to get around similar issues.

Ideally, conky should quit if X11 isn't available. But if something like display variables are missing, it should busy-wait until they're ready (as X11 backend assumes some graphics will be shown).

**Comment by rochfort1** [comment]
I don't think that this can be called a niche bug. The workarounds are clunky to say the least.

Some users report that they have added the following lines to .conkyrc at the top of the conky.config section:

conky.config = {
    out_to_x = false,
    out_to_wayland = true,

These users report that conky starts and operates normally under Wayland without first having to start Xwayland. However if I set up conky.config with those lines and then run $ conky, I see the following:

GNOME 47 under Debian 13 and Arch Linux
conky: drawing to single buffer
conky: FOUND: console
conky: FOUND: ncurses
conky: FOUND: file
conky: FOUND: x11
conky: FOUND: wayland
conky: Compositor doesn't support wlr-layer-shell-unstable-v1. Can't run conky.

Fedora 41
conky: drawing to single buffer
conky: Unknown setting 'out_to_wayland'
conky: FOUND: console
conky: FOUND: ncurses
conky: FOUND: file
conky: FOUND: x11
Unable to find a usable display output.
conky: 'gnome' wayland session running 'GNOME' desktop

It's curious that conky reports that it found x11. System Monitor does not list XWayland.

Versions of conky currently installed are:
Debian 13: conky-std 1.21.9-1
Arch Linux: conky 1.21.5-1
Fedora 41: conky.x86_64 1.21.7-1.fc41

**Comment by Caellian** [comment]
> I don't think that this can be called a niche bug. The workarounds are clunky to say the least.

Wayland support is spotty, if you use XWayland conky is forced to use XDG shell by most compositors which means it will show up as a regular window and that's not something most people want. So _that_ is a niche use case.

Wayland (not [XWayland](https://wiki.archlinux.org/title/Wayland#Xwayland)) support is being worked on, but that's unrelated to this issue. See [wayland related issues](https://github.com/brndnmtthws/conky/issues?q=is%3Aopen+is%3Aissue+label%3A%22display%3A+wayland%22+-label%3A%22env%3A+xwayland%22).

With #2103 merged, Wayland has higher priority than XWayland, so this issue is for the case when people run conky on Wayland using:
```
out_to_x=true,
out_to_wayland=false,
```

> It's curious that conky reports that it found x11. System Monitor does not list XWayland.

It didn't find X11, it found an output for X11, i.e. part of code that allows it to draw on X11 was compiled into the application.
